### ISSUE TITLE
USB Drive dismount problem

### ISSUE THREAD

**Issue by wards999** [bailey27/cppcryptfs#126]
1. I created and mounted a new cppcryptfs folder on a usb drive and started copying files to it using FreeFileSync.
2. After the sync was complete I dismounted the drive from within cppcryptfs.
3. When I tried to dismount the usb drive from Windows, I got the message 'the device is currently in use'
4. Using Process Explorer I found that cppcryptfs.exe still had handles to 356 files.
5. After exiting cppcryptfs.exe completely I could then dismount the usb drive successfully.

Does this mean that there might have been data loss on the files referenced by the handles when the exe was closed? 
Process Explorer lists the encrypted folder/filenames, but I can't (obviously) tie that back to the unencrypted name to check.

Just like Windows warns against unmounting, shouldn't cppcryptfs also warn the user rather than allowing the dismount to go ahead, if there are open handles?

Thanks.

**Comment by bailey27** [comment]
I tried this, and what I see is that cppcryptfs is leaking (losing track of) handles to the root of the unencrypted fs (to the directory).  It looks like it leaks two handles like this each time an fs is mounted.

e.g. if I make a filesystem on my usb drive (D:) in d:\test, then after syncing some files with FreeFileSync (or just copying with file explorer) to the mounted fs (e.g. mounted on k:\), then if I unmount it I can see in process explorer that there are still two handles open to d:\test (the directory itself).  All the other handles that are open by cppcryptfs appear to be for things like DLLs, registry keys, threads, semaphores, sections, etc.  The other open handles look ok to me.

The USB drive can't be ejected because of the open handles to d:\test in my case.  If I close those handles in process explorer, then I can eject the usb drive.

This is definitely a bug that should be fixed, but I think there is no risk of data loss because of it.

**Comment by wards999** [comment]
Thanks very much for the reply, although if I understand correctly, it seems your test showed only 2  handles of significance that needed to be closed. 

However, when I looked at the handles associated with the external drive letter via Process Explorer on my machine, all 356 appeared to be pointing to encrypted files. I'm going to run another FreeFileSync session with 1000s of files across multiple directories and again check the results, but good to know that there shouldn't be any data loss.

I don't know if my accessing some of the folders / files that had already been copied, while FFS was continuing to run in the background might have contributed to the issue?

**Comment by wards999** [comment]
Just a quick update:

I've run 3 more large FreeFileSync sessions using a usb cppcryptfs target. 

The first two left no handles and so the usb drive was able to be dismounted. The third left 2 handles to the directory on the usb drive, exactly the same as your result. Once these were removed the drive was able to be dismounted.

(I'm still puzzled how my initial experience with 356 handles to encrypted files occurred, but for now I cannot reproduce it)

**Comment by wards999** [comment]
And another:

Was running a winmerge diff check against the cppcryptfs files and after dismounting it's showing 146 handles. 
2 are for the directory - the rest of pointing at encrypted files...

**Comment by bailey27** [comment]
I did some testing with Winmerge and an encrypted fs on a USB drive.

I was not able to reproduce the mass handle leakage that you were experiencing.

What I did see is that if an app opens a handle and keeps it open, then cppcryptfs never gets told to close it if the fs is dismounted.  So the handle stays open and is leaked by cppcryptfs.  I was able to reproduce this with a simple test program that opens a file and then goes in an infinite sleep for 100ms loop.

I also determined that those handles to the root of the encrypted fs I saw being leaked before are not being opened by cppcryptfs internally.  An external process is opening a handle to the root of the unencrypted view, and cppcryptfs is opening a companion handle to the root of the encrypted fs, and cppcryptfs is never being told to close it.

The way things work is that when an unencrypted file/directory is opened, cppcryptfs stores its handle to the counterpart encrypted file/directory  in the file structure passed to it by Dokany.  And when whoever opened the unencrypted file/directory  closes it, then cppcryptfs gets the request from dokany and closes the handle to the encrypted counterpart that it saved in the Dokany file structure.

Apparently, when the fs is dismounted, Dokany doesn't generate any requests to clean up these handles that are still open when the fs is dismounted.

I think the best solution would be for cppcryptfs to keep track of all the open file handles and close any that remain open when the fs is dismounted.

I think I should also add a setting to warn the user if there are open handles on the fs when the user goes to dismount it.
